drivers: uart: Refactor nrf uart shim

jakub-uC opened the pull request:
Replace all register defines and calls with Nordic nrfx HAL.

Fixes #7755 

Signed-off-by: Jakub Rzeszutko <Jakub.Rzeszutko@nordicsemi.no>

Commit:
drivers: uart: Refactor nrf uart shim

Replace all register defines and calls with Nordic nrfx HAL.
Simplification of uart shim - assumed only one uart instance.
Added parity bit to configuration options.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

carlescufi commented on arch/arm/soc/nordic_nrf/include/nrf_common.h:
```diff
@@ -61,30 +61,4 @@
- * Uses the UARTE_BAUDRATE macros since they are more precise.
- */
-
-#define NRF_UART_BAUDRATE_300                  0x00014000
```
The reason that I added those is that the ones in the MDK for UART were not very accurate, so i used the ones for UARTE

carlescufi replied:
Can you verify that, with the refactor, these same values will be used?

jarz-nordic replied:
I do not want to mix these defines in uart_nrf5.c file. I will use UARTE defines in new shim for uarte.
In m opinion inaccuracy in UART defines is not so signifacant to cause any problem during the serial transsmition. We do not see complains in SDK releases about this.

carlescufi commented on drivers/serial/uart_nrf5.c:
```diff
@@ -305,13 +250,13 @@ static unsigned char uart_nrf5_poll_out(struct device *dev,
 
 	/* reset transmitter ready state */
-	uart->EVENTS_TXDRDY = 0;
+	nrf_uart_event_clear(uart, NRF_UART_EVENT_TXDRDY);
```
This has been recently changed upstream to be done **after** the `while` loop due to a race condition. Can you take a look?

carlescufi replied:
I am talking about [this change](https://github.com/zephyrproject-rtos/zephyr/commit/05c45e359a94e96353dfac1f0abeccf6893e988a#diff-fb209e20d9cf9e70b3dbd2748ae43f90) see however issue #7755, which seems to have been introduced by it. Hence I believe we need to incorporate that fix but also resolve the issue it introduces.

lemrey replied:
It seems that the change you mentioned is included in this PR: the register used to be cleared at the end of the function, now it is cleared before. That seems to have broken the console in master (see issue [#7755](https://github.com/zephyrproject-rtos/zephyr/issues/7755)) but it seems to be working fine in this PR.

carlescufi replied:
Ah true @lemrey I misread the code.

anangl commented on drivers/serial/uart_nrf5.c:
```diff
@@ -221,29 +162,34 @@ static int uart_nrf5_init(struct device *dev)
 	/* Set baud rate */
-	err = baudrate_set(dev, DEV_DATA(dev)->baud_rate,
-		     DEV_CFG(dev)->sys_clk_freq);
+	err = baudrate_set(dev, DEV_DATA(dev)->baud_rate);
```
It seems pointless to dedicate the `baud_rate` field for this. You can use `CONFIG_UART_NRF5_BAUD_RATE` directly here.

anangl commented on drivers/serial/uart_nrf5.c:
```diff
@@ -221,29 +162,34 @@ static int uart_nrf5_init(struct device *dev)
+	nrf_uart_task_trigger(uart, NRF_UART_TASK_STARTTX);
+	nrf_uart_task_trigger(uart, NRF_UART_TASK_STARTRX);
 
 	dev->driver_api = &uart_nrf5_driver_api;
```
This should be done by the `DEVICE_AND_API_INIT` macro which should be used instead of `DEVICE_INIT` at the end of this file.

anangl commented on drivers/serial/uart_nrf5.c:
```diff
@@ -196,7 +136,7 @@ static int baudrate_set(struct device *dev,
 {
-	volatile struct _uart *uart = UART_STRUCT(dev);
+	NRF_UART_Type *uart = UART_STRUCT(dev);
 	struct device *gpio_dev;
```
Binding of this device might fail but this is not checked (and should be at least with an assertion).
I can't place a comment on line 143.

anangl commented:
In a separate commit `uart_nrf5.c` should be renamed to `uart_nrfx_uart.c` (as `uart_nrfx_uarte.c` is supposed to appear soon). I think it would be good to make it part of this PR.
Occurences of "nrf5" and "NRF5" in names of functions, Kconfig entries etc. should be updated, too (to "nrfx" or "nrf", accordingly).

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -17,53 +17,61 @@ menuconfig UART_NRF5
 	range 0 31
-	depends on UART_NRF5
+	depends on UART_NRFX_UART
 	depends on GPIO_NRF5
```
This is unnecessary. The `_TX_PIN` option only allows to select a pin number, it does not need GPIO_NRF5 to do so.
This applies to the other pin configs here as well.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,443 @@
+ * @param baudrate Baud rate
+ * @param sys_clk_freq_hz System clock frequency in Hz
+ *
+ * @return N/A
```
This description is outdated (parameters and return value).

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,443 @@
+ * @return N/A
+ */
+
+static int baudrate_set(struct device *dev)
```
The `baudrate` parameter should stay. Without it `baudrate_set` looks weird.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,443 @@
+{
+	nrf_uart_baudrate_t nrf_baudrate; /* calculated baudrate divisor */
+
+	/* Use the common nRF5 macros */
```
I don't really understand what the comment is telling me. But for sure, this "5" should be gone.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,443 @@
+ * @return 0 if a character arrived, -1 if the input buffer if empty.
+ */
+
+static inline int uart_nrf5_poll_in(struct device *dev, unsigned char *c)
```
Again, "nrf5" should no longer occur. Below as well.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,443 @@
+#ifdef CONFIG_UART_NRFX_UART_PARITY_BIT
+	nrf_uart_configure(NRF_UART0,
+			   NRF_UART_PARITY_INCLUDED,
+			   NRF_UART_HWFC_ENABLED);
```
```
	nrf_uart_configure(NRF_UART0,
#ifdef CONFIG_..._PARITY_BIT
			   NRF_UART_PARITY_INCLUDED,
#else
			   NRF_UART_PARITY_EXCLUDED,
#endif
#ifdef CONFIG_..._FLOW_CONTROL
			   NRF_UART_HWFC_ENABLED);
#else
			   NRF_UART_HWFC_DISABLED);
#endif
```
Less error-prone.

jarz-nordic replied:
good point!

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,443 @@
+ *
+ * @return 0 on success
+ */
+static int uart_nrfx_uart_init(struct device *dev)
```
`uart_nrfx_init` should be sufficient.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,443 @@
+
+	gpio_dev = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	__ASSERT(gpio_dev, "UART binding failed");
```
"Cannot find %s", CONFIG_GPIO_NRF5_P0_DEV_NAME).

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -5,7 +5,7 @@
 # SPDX-License-Identifier: Apache-2.0
```
This header needs to be updated ("nrf5"),

jarz-nordic commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -17,53 +17,61 @@ menuconfig UART_NRF5
-config UART_NRF5_FLOW_CONTROL
+config UART_NRFX_UART_PARITY_BIT
+        bool "Parity bit enabled"
+        default n
```
Ok,
The same I will remove for UART_NRFX_FLOW_CONTROL

jarz-nordic commented:
Code has been updated according to your comments.

anangl commented on arch/arm/soc/nordic_nrf/nrf51/dts.fixup:
```diff
@@ -1,9 +1,9 @@
-#define CONFIG_UART_NRF5_NAME		NORDIC_NRF_UART_40002000_LABEL
+#define CONFIG_UART_NRFX_IRQ_PRI	NORDIC_NRF_UART_40002000_IRQ_0_PRIORITY
+#define CONFIG_UART_NRFX_BAUD_RATE	NORDIC_NRF_UART_40002000_CURRENT_SPEED
+#define CONFIG_UART_NRFX_NAME		NORDIC_NRF_UART_40002000_LABEL
```
`#define CONFIG_UART_0_IRQ_PRI` and so on.

jarz-nordic replied:
fixed

anangl commented on samples/application_development/out_of_tree_board/boards/arm/nrf52840_pca10056/nrf52840_pca10056_defconfig:
```diff
@@ -9,7 +9,7 @@ CONFIG_ARM_MPU_NRF52X=y
 CONFIG_SERIAL=y
-CONFIG_UART_NRF5=y
+CONFIG_UART_NRFX=y
 
```
`CONFIG_UART_0_NRF_UART=y`

jarz-nordic replied:
fixed

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -0,0 +1,79 @@
+          Enable parity bit.
+
+config UART_0_NRF_FLOW_CONTROL
+	bool "Enable Flow Control"
```
"Enable flow control"

jarz-nordic replied:
fixed

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -0,0 +1,79 @@
+if UART_0_NRF_UART
+
+config UART_0_NRF_PARITY_BIT
+        bool "Parity bit enabled"
```
"Enable parity bit", to be consistent with "Enable flow control".

jarz-nordic replied:
fixed

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -0,0 +1,79 @@
+	bool "Enable Flow Control"
+	help
+	  Enable Flow Control. If selected, additionally two pins, RTS and CTS
+	  has to be configured.
```
have

jarz-nordic replied:
fixed

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -0,0 +1,79 @@
+config UART_0_NRF_UART
+	bool "nRF UART 0"
+	depends on ((SOC_SERIES_NRF52X || SOC_SERIES_NRF51X) && (!SOC_NRF52810))
+    select NRF_UART_PERIPHERAL
```
This indentation should be corrected.

jarz-nordic replied:
fixed

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -0,0 +1,79 @@
+# Kconfig.nrfx - NRFX UART configuration
```
nrfx

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,419 @@
+
+	__ASSERT(gpio_dev,
+			 "UART init failed. Cannot find %s",
+			 CONFIG_GPIO_NRF5_P0_DEV_NAME);
```
This indentation is a bit weird. One TAB too far?

jarz-nordic replied:
I used 4 character tab instead of 8, I will fix it.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,419 @@
+
+	nrf_uart_configure(NRF_UART0,
+#ifdef CONFIG_UART_0_NRF_PARITY_BIT
+					   NRF_UART_PARITY_INCLUDED,
```
again, weird indentation

jarz-nordic replied:
fixed

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -0,0 +1,79 @@
+endif # UART_NRFX
+
+config NRF_UART_PERIPHERAL
+    bool "When set uart_nrfx_uart.c will be added to the project."
```
This should be a hidden setting (not to be changed by a user directly). By adding a prompt you make it visible, in menuconfig for instance.

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -0,0 +1,79 @@
+endif # UART_NRFX
+
+config NRF_UART_PERIPHERAL
+	bool "When set uart_nrfx_uart.c will be added to the project."
```
This entry should not contain a prompt (see [my previous comment](https://github.com/zephyrproject-rtos/zephyr/pull/7681#pullrequestreview-126702201)).

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -24,17 +24,15 @@ static uart_irq_callback_t m_irq_callback; /**< Callback function pointer */
  *
  * @param dev UART device struct
  * @param baudrate Baud rate
- * @param sys_clk_freq_hz System clock frequency in Hz
```
This change (as well as a few more below) should not be contained in this commit as it has nothing to do with the nrf5 renaming. Move all the changes related to shim refactoring to the first commit.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,419 @@
+ * @return 0 if a character arrived, -1 if the input buffer if empty.
+ */
+
+static inline int uart_nrfx_poll_in(struct device *dev, unsigned char *c)
```
All these functions cannot be "inline", since you fill the `uart_driver_api` structure with pointers to them. Some compilers might complain about it. Besides, it may confuse the ones reading the source code.

anangl commented:
@jarz Commit messages needs to be updated.
The first one does not mention the addition of parity bit support nor the simplification of the shim based on the fact that there is only one instance of the UART.
The second one does not really describe the actual changes made.

jarz-nordic commented on arch/arm/soc/nordic_nrf/nrf51/dts.fixup:
```diff
@@ -1,9 +1,9 @@
-#define CONFIG_UART_NRF5_BAUD_RATE	NORDIC_NRF_UART_40002000_CURRENT_SPEED
-#define CONFIG_UART_NRF5_NAME		NORDIC_NRF_UART_40002000_LABEL
+#define CONFIG_UART_0_IRQ_PRI	NORDIC_NRF_UART_40002000_IRQ_0_PRIORITY
+#define CONFIG_UART_0_BAUD_RATE	NORDIC_NRF_UART_40002000_CURRENT_SPEED
```
done

anangl replied:
It looks like it is not.
Maybe something eats your TABs? Some starving process, a philosopher? ;-)

jarz-nordic commented:
I've reworked commits.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -325,54 +324,42 @@ DEVICE_DECLARE(uart_nrfx_uart0);
-	__ASSERT(gpio_dev, "UART binding failed");
+	__ASSERT(gpio_dev,
+		 "UART init failed. Cannot find %s",
+		 CONFIG_GPIO_NRF5_P0_DEV_NAME);
```
This change should also be moved to the first commit.

anangl commented on drivers/serial/uart_nrf5.c:
```diff
@@ -263,19 +108,17 @@ static int uart_nrf5_init(struct device *dev)
  */
 
-static int uart_nrf5_poll_in(struct device *dev, unsigned char *c)
+static inline int uart_nrf5_poll_in(struct device *dev, unsigned char *c)
```
These "inline" keywords were supposed to disappear completely. Three of them still appear in this commit and are removed in the second one.

Commit:
drivers: uart: Rename nrf5 namings to nrfx

UART driver renamed to keep the same convention as SPI and TWI drivers.
All substrings: "UART_NRF5" in defines renamed to  "UART_NRFX_UART".

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

jarz-nordic commented:
I apologize, it should be good now.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -0,0 +1,419 @@
+ * @return Sent character
+ */
+static unsigned char uart_nrfx_poll_out(struct device *dev,
+					       unsigned char c)
```
Almost there. Just correct this indentation and it will be fine.

jarz-nordic commented:
done
